### PR TITLE
Bumping Swarmkit's version

### PR DESCRIPTION
**Pull request by wk8**
Mainly for the sake of testing the new "soft delete" capabilities of
Swarmkit, see more at
https://github.com/docker/swarmkit/pull/2778

Also reflecting new Swarmkit's API changes in moby's own types, namely
the `PendingDelete` fields added to both `Services` and `NetworkResources`.

Also amended relevant tests to now wait for services to be fully shut down
after removing them.

Signed-off-by: Jean Rouge <jean.rouge@docker.com>

### CODE REVIEW

**Comment by olljanat** [comment]
@wk8 Janky failure:
```
02:50:56 ============================= test session starts ==============================
02:50:56 platform linux2 -- Python 2.7.13, pytest-2.9.1, py-1.7.0, pluggy-0.3.1
02:50:56 rootdir: /docker-py, inifile: pytest.ini
02:50:56 plugins: cov-2.1.0
02:50:57 collected 338 items
02:50:57 
02:51:23 ../../../../../docker-py/tests/integration/api_build_test.py ....s..s........
02:51:24 ../../../../../docker-py/tests/integration/api_client_test.py .......
02:51:25 ../../../../../docker-py/tests/integration/api_config_test.py .....
02:52:49 ../../../../../docker-py/tests/integration/api_container_test.py ....x....x.......x............................................................
02:52:59 ../../../../../docker-py/tests/integration/api_exec_test.py .............
02:53:06 ../../../../../docker-py/tests/integration/api_healthcheck_test.py ....
02:53:17 ../../../../../docker-py/tests/integration/api_image_test.py ...s........s....
02:53:52 ../../../../../docker-py/tests/integration/api_network_test.py ...........................
02:54:15 ../../../../../docker-py/tests/integration/api_plugin_test.py ...........
02:54:16 ../../../../../docker-py/tests/integration/api_secret_test.py .....
02:54:35 ../../../../../docker-py/tests/integration/api_service_test.py ............................FF.......F.FF.............
02:55:07 ../../../../../docker-py/tests/integration/api_swarm_test.py ..s...x......
02:55:22 ../../../../../docker-py/tests/integration/api_volume_test.py .........
02:55:22 ../../../../../docker-py/tests/integration/client_test.py ....
02:55:23 ../../../../../docker-py/tests/integration/errors_test.py .
02:56:11 ../../../../../docker-py/tests/integration/models_containers_test.py ..............................
02:56:27 ../../../../../docker-py/tests/integration/models_images_test.py .............
02:56:28 ../../../../../docker-py/tests/integration/models_networks_test.py ....
02:56:34 ../../../../../docker-py/tests/integration/models_nodes_test.py .
02:56:35 ../../../../../docker-py/tests/integration/models_resources_test.py .
02:56:37 ../../../../../docker-py/tests/integration/models_services_test.py ......F.......X.
02:56:38 ../../../../../docker-py/tests/integration/models_swarm_test.py .
02:56:38 ../../../../../docker-py/tests/integration/models_volumes_test.py ..
02:56:42 ../../../../../docker-py/tests/integration/regression_test.py ......
02:56:42 
02:56:42  generated xml file: /go/src/github.com/docker/docker/bundles/test-docker-py/results.xml 
02:56:42 =========================== short test summary info ============================
02:56:42 XFAIL ../../../../../docker-py/tests/integration/api_container_test.py::CreateContainerTest::test_create_with_cpu_rt_options
02:56:42   CONFIG_RT_GROUP_SCHED isn't enabled
02:56:42 XFAIL ../../../../../docker-py/tests/integration/api_container_test.py::CreateContainerTest::test_create_with_init_path
02:56:42   init-path removed in 17.05.0
02:56:42 XFAIL ../../../../../docker-py/tests/integration/api_container_test.py::CreateContainerTest::test_create_with_storage_opt
02:56:42   Not supported on most drivers
02:56:42 XFAIL ../../../../../docker-py/tests/integration/api_swarm_test.py::SwarmTest::test_init_swarm_with_log_driver
02:56:42   This doesn't seem to be taken into account by the engine
02:56:42 SKIP [1] tests/integration/api_image_test.py:289: Doesn't work inside a container - FIXME
02:56:42 SKIP [1] /docker-py/tests/integration/api_swarm_test.py:31: Test stalls the engine on 1.12.0
02:56:42 SKIP [3] /docker-py/tests/helpers.py:69: Feature requires Docker Engine experimental mode
02:56:42 =================================== FAILURES ===================================
02:56:42 _____________ ServiceTest.test_create_service_with_unicode_config ______________
02:56:42 /docker-py/tests/integration/api_service_test.py:642: in test_create_service_with_unicode_config
02:56:42     config_id = self.client.create_config(config_name, config_data)
02:56:42 /docker-py/docker/utils/decorators.py:34: in wrapper
02:56:42     return f(self, *args, **kwargs)
02:56:42 /docker-py/docker/api/config.py:35: in create_config
02:56:42     self._post_json(url, data=body), True
02:56:42 /docker-py/docker/api/client.py:229: in _result
02:56:42     self._raise_for_status(response)
02:56:42 /docker-py/docker/api/client.py:225: in _raise_for_status
02:56:42     raise create_api_error_from_http_exception(e)
02:56:42 /docker-py/docker/errors.py:31: in create_api_error_from_http_exception
02:56:42     raise cls(e, response=response, explanation=explanation)
02:56:42 E   APIError: 409 Client Error: Conflict ("rpc error: code = AlreadyExists desc = config favorite_touhou already exists")
02:56:42 _____________ ServiceTest.test_create_service_with_unicode_secret ______________
02:56:42 /docker-py/tests/integration/api_service_test.py:590: in test_create_service_with_unicode_secret
02:56:42     secret_id = self.client.create_secret(secret_name, secret_data)
02:56:42 /docker-py/docker/utils/decorators.py:34: in wrapper
02:56:42     return f(self, *args, **kwargs)
02:56:42 /docker-py/docker/api/secret.py:46: in create_secret
02:56:42     self._post_json(url, data=body), True
02:56:42 /docker-py/docker/api/client.py:229: in _result
02:56:42     self._raise_for_status(response)
02:56:42 /docker-py/docker/api/client.py:225: in _raise_for_status
02:56:42     raise create_api_error_from_http_exception(e)
02:56:42 /docker-py/docker/errors.py:31: in create_api_error_from_http_exception
02:56:42     raise cls(e, response=response, explanation=explanation)
02:56:42 E   APIError: 409 Client Error: Conflict ("rpc error: code = AlreadyExists desc = secret favorite_touhou already exists")
02:56:42 ________________________ ServiceTest.test_list_services ________________________
02:56:42 /docker-py/tests/integration/api_service_test.py:76: in test_list_services
02:56:42     assert len(test_services) == 0
02:56:42 E   AssertionError: assert 9 == 0
02:56:42 E    +  where 9 = len([{'CreatedAt': '2019-01-10T02:54:17.610899445Z', 'Endpoint': {'Spec': {}}, 'ID': 'obpnh07dag0cwi8z2yu3phd68', 'Pending...10T02:54:23.874616888Z', 'Endpoint': {'Spec': {}}, 'ID': 'x1i3jvdrgrei9h1yj9i9l2bk7', 'PendingDelete': True, ...}, ...])
02:56:42 ____________________ ServiceTest.test_remove_service_by_id _____________________
02:56:42 /docker-py/tests/integration/api_service_test.py:118: in test_remove_service_by_id
02:56:42     assert len(test_services) == 0
02:56:42 E   AssertionError: assert 4 == 0
02:56:42 E    +  where 4 = len([{'CreatedAt': '2019-01-10T02:54:17.610899445Z', 'Endpoint': {'Spec': {}}, 'ID': 'obpnh07dag0cwi8z2yu3phd68', 'Pending...9-01-10T02:54:24.061397702Z', 'Endpoint': {'Spec': {}}, 'ID': 'fet1grsijuqh9q06mo4t7t8w7', 'PendingDelete': True, ...}])
02:56:42 ___________________ ServiceTest.test_remove_service_by_name ____________________
02:56:42 /docker-py/tests/integration/api_service_test.py:124: in test_remove_service_by_name
02:56:42     assert len(test_services) == 0
02:56:42 E   AssertionError: assert 5 == 0
02:56:42 E    +  where 5 = len([{'CreatedAt': '2019-01-10T02:54:24.142811361Z', 'Endpoint': {'Spec': {}}, 'ID': '0p7bqsqb23lezynqqaquroex0', 'Pending...9-01-10T02:54:24.061397702Z', 'Endpoint': {'Spec': {}}, 'ID': 'fet1grsijuqh9q06mo4t7t8w7', 'PendingDelete': True, ...}])
02:56:42 _________________________ ServiceTest.test_list_remove _________________________
02:56:42 /docker-py/tests/integration/models_services_test.py:80: in test_list_remove
02:56:42     assert service not in client.services.list()
02:56:42 E   AssertionError: assert <Service: bke7s3fwnr> not in [<Service: bke7s3fwnr>, <Service: de3pe7myoh>, <Service: df7yub3t6n>, <Service: ozisoi19iv>, <Service: u4fwx2vhug>, <Service: uwsemomnq7>, ...]
02:56:42 E    +  where [<Service: bke7s3fwnr>, <Service: de3pe7myoh>, <Service: df7yub3t6n>, <Service: ozisoi19iv>, <Service: u4fwx2vhug>, <Service: uwsemomnq7>, ...] = <bound method ServiceCollection.list of <docker.models.services.ServiceCollection object at 0x7fd67c4c6510>>()
02:56:42 E    +    where <bound method ServiceCollection.list of <docker.models.services.ServiceCollection object at 0x7fd67c4c6510>> = <docker.models.services.ServiceCollection object at 0x7fd67c4c6510>.list
02:56:42 E    +      where <docker.models.services.ServiceCollection object at 0x7fd67c4c6510> = <docker.client.DockerClient object at 0x7fd67c4d80d0>.services
02:56:42 === 6 failed, 322 passed, 5 skipped, 4 xfailed, 1 xpassed in 346.28 seconds ====
06:00:50 Build timed out (after 200 minutes). Marking the build as failed.
06:00:50 Build timed out (after 200 minutes). Marking the build as aborted.
```

And Experimental failure:
```
03:47:46 ----------------------------------------------------------------------
03:47:46 FAIL: docker_api_swarm_service_test.go:402: DockerSwarmSuite.TestAPISwarmServiceConstraintLabel
03:47:46 
03:47:46 [d19a4a7199ab1] waiting for daemon to start
03:47:46 [d19a4a7199ab1] daemon started
03:47:46 
03:47:46 [da3902fd25507] waiting for daemon to start
03:47:46 [da3902fd25507] daemon started
03:47:46 
03:47:46 [dc49c2f95d361] waiting for daemon to start
03:47:46 [dc49c2f95d361] daemon started
03:47:46 
03:47:46 waited for 1.376717ms (out of 5s)
03:47:46 waited for 924.416453ms (out of 30s)
03:47:46 assertion failed: error is not nil: Error response from daemon: rpc error: code = AlreadyExists desc = name conflicts with an existing object: service top already exists
03:47:46 [d19a4a7199ab1] exiting daemon
03:47:46 [da3902fd25507] exiting daemon
03:47:46 [dc49c2f95d361] exiting daemon
03:48:01 
03:48:01 ----------------------------------------------------------------------
03:48:01 FAIL: docker_api_swarm_service_test.go:350: DockerSwarmSuite.TestAPISwarmServiceConstraintRole
03:48:01 
03:48:01 [d0b43fa1e3376] waiting for daemon to start
03:48:01 [d0b43fa1e3376] daemon started
03:48:01 
03:48:01 [dfecb3662b28b] waiting for daemon to start
03:48:01 [dfecb3662b28b] daemon started
03:48:01 
03:48:01 [da65e7126d230] waiting for daemon to start
03:48:01 [da65e7126d230] daemon started
03:48:01 
03:48:01 waited for 1.398944ms (out of 5s)
03:48:01 waited for 927.794596ms (out of 30s)
03:48:01 assertion failed: error is not nil: Error response from daemon: rpc error: code = AlreadyExists desc = name conflicts with an existing object: service top already exists
03:48:01 [d0b43fa1e3376] exiting daemon
03:48:01 [dfecb3662b28b] exiting daemon
03:48:01 [da65e7126d230] exiting daemon
03:48:16 
```

Looks to be new issues which at least I have not seen earlier.


Z looks to be infra issue `java.nio.file.DirectoryNotEmptyException`
and PowerPC failed to #38460 so I will trigger rebuild for those.

**Comment by olljanat** [comment]
Derek add label: rebuild/powerpc

**Comment by cpuguy83** [comment]
Yeah, this seems like a breaking change from swarmkit.

**Comment by wk8** [comment]
Note to reviewers: the build failure should be fixed once https://github.com/docker/docker-py/pull/2228 is merged. Thanks!

File: internal/test/daemon/config.go [hunk 1]
```diff
@@ -14,9 +13,6 @@ type ConfigConstructor func(*swarm.Config)
 
 // CreateConfig creates a config given the specified spec
 func (d *Daemon) CreateConfig(t assert.TestingT, configSpec swarm.ConfigSpec) string {
-	if ht, ok := t.(test.HelperT); ok {
-		ht.Helper()
-	}
```
Comment (wk8) [inline internal/test/daemon/config.go]: Note for all of these: `d.NewClientT(t)` already does the exact same thing.

**Comment by cpuguy83** [comment]
Sorry no, changing docker-py is not a valid fix for a breaking API change.

**Comment by cpuguy83** [comment]
These tests exist to pinpoint issues like this, just changing the test is not how to deal with this.

This is a breaking change in swarmkit. It should be addressed there, and if not then we need to change the implementation here such that these changes are not breaking existing users.

**Comment by wk8** [comment]
@cpuguy83 : yes, this is absolutely a breaking change in Swarmkit, that has been deemed the lesser of two evils after ample discussion within the Swarmkit team and outside of it.

Since it is indeed a breaking change, and changes a behavior that tests both in this repo and in docker-py rely upon, it seems logical to me to adapt the tests to test the new desired behavior?

**Comment by olljanat** [comment]
@wk8 remember that Moby API must be backward compatible so you cannot change behavior for API versions < 1.40

**Comment by cpuguy83** [comment]
Glossing over the actual bug being solved... I'll assume this had due diligence applied in how to solve the problem... why can't async delete be opt-in?

If swarmkit is not going to handle this, then we need to handle it in moby. We simply cannot have such a breaking change.

Basically:
```
Create("foo")
Delete("foo")
Create("foo")
```

Should just work. If `Delete` returns then "foo" must be available immediately again.

**Comment by cpuguy83** [comment]
btw, when I say "opt-in", I mean the delete internally would still be async (marked for removal) but the API would either only return when it is actually freed or if async=true it can return as soon as it has been marked.

**Comment by wk8** [comment]
@cpuguy83 : sorry, I certainly didn't mean to gloss over the actual problem, apologies if it came across that way.

The actual problem being solved is that currently nothing prevents you from creating a service, with a network attached to it, and then deleting the service and the network even if the service's containers have not shut down yet. This means the containers can't perform their cleanup properly if they need to use the network to do so; and can also lead to interesting failures if a network and/or a service with the same name are re-created before the old ones are gone.

The same applies to any service-level resource down the road.

Note that it's only a breaking change precisely if you do anything that assumes (and _needs_ to assume) that a service gets instantaneously deleted, which is conceptually wrong, and will lead to errors if that's really an assumption you need.

I understand your concern, though; how would you feel about, depending on the client's version:
* if < 1.40, then we only return from the API call after the delete has occurred
* otherwise, we return instantly, as we used to, but the actual removal happens asynchronously

That would ensure we don't break older clients, while still also returning fast in the 99% use case, when people don't try re-creating a service or a service-level resource right away.

**Review by thaJeztah — CHANGES_REQUESTED** [review]
See my comment inline; about adding a separate field for this.

I also agree with @cpuguy83  - we can't change this behavior and make it the default; doing so will be breaking any consumer of the API out there; this behavior should be opt-in, and enabled through a new option in the API request (to be added in API v1.40 and up).

File: api/types/swarm/service.go [hunk 2]
```diff
@@ -6,10 +6,11 @@ import "time"
...
 	Meta
-	Spec         ServiceSpec   `json:",omitempty"`
-	PreviousSpec *ServiceSpec  `json:",omitempty"`
-	Endpoint     Endpoint      `json:",omitempty"`
-	UpdateStatus *UpdateStatus `json:",omitempty"`
+	Spec          ServiceSpec   `json:",omitempty"`
+	PreviousSpec  *ServiceSpec  `json:",omitempty"`
+	Endpoint      Endpoint      `json:",omitempty"`
+	UpdateStatus  *UpdateStatus `json:",omitempty"`
+	PendingDelete bool          `json:",omitempty"`
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply api/types/swarm/service.go]: K8s uses `DeletionTimestap`, though it is a little different, basically the timestamp is the time when the object should no longer be visible:

> The resource is expected to be deleted (no longer visible
    // from resource lists, and not reachable by name) after the time in this field, once the
    // finalizers list is empty.

**Comment by thaJeztah** [comment]
Closing for now

**Pull request closed without merging**